transport/src/main/java/io/servicefabric/transport/Transport.java

```diff
@@ -164,72 +174,61 @@ protected void initChannel(Channel channel) {
-      LOGGER.info("Netty TransportFactory - bound to: {}", bindAddress);
+      LOGGER.info("Transport endpoint '{}' bound to: {}", localEndpoint.id(), bindAddress);
     } catch (Exception e) {
       LOGGER.error("Failed to bind to: " + bindAddress + ", caught " + e, e);
       propagate(e);
     }
   }
 
-  public EventLoopGroup getEventLoop() {
-    return eventLoop;
-  }
-
   @Override
-  public EventExecutorGroup getEventExecutor() {
-    return eventExecutor;
+  public ListenableFuture<TransportEndpoint> connect(@CheckForNull final TransportAddress address) {
+    checkArgument(address != null);
+    final TransportChannel transportChannel = getOrConnect(address);
+    return Futures.transform(transportChannel.handshakeFuture(),
+        new AsyncFunction<TransportHandshakeData, TransportEndpoint>() {
+          @Override
+          public ListenableFuture<TransportEndpoint> apply(@Nonnull TransportHandshakeData handshakeData)
+              throws Exception {
+            return Futures.immediateFuture(handshakeData.endpoint());
+          }
+        });
   }
 
-  @Nonnull
   @Override
-  public final ITransportChannel to(@CheckForNull final TransportEndpoint endpoint) {
+  public void disconnect(@CheckForNull TransportEndpoint endpoint, @Nullable SettableFuture<Void> promise) {
     checkArgument(endpoint != null);
-    return connected.get(endpoint, new Computable<TransportEndpoint, TransportChannel>() {
-      @Override
-      public TransportChannel compute(TransportEndpoint arg) {
-        final Channel channel = createConnectorChannel();
-        final TransportChannel transport = createConnector(channel, endpoint);
-
-        channel.attr(ATTR_TRANSPORT).set(transport);
-        LOGGER.debug("Registered connector: {}", transport);
-
-        final ChannelFuture regFuture = eventLoop.register(channel);
-        if (regFuture.cause() != null) {
-          if (channel.isRegistered()) {
-            channel.close();
-          } else {
-            channel.unsafe().closeForcibly();
-          }
-          throw new TransportException(transport, regFuture.cause());
-        }
+    ITransportChannel transportChannel = connectedChannels.getIfExists(endpoint.address());
+    // TODO [AK]: check that channel endpoint id correspond to provided endpoint id; fail otherwise
+    if (transportChannel == null) {
+      if (promise != null) {
+        promise.set(null);
+      }
+    } else {
+      transportChannel.close(promise);
+    }
+  }
 
-        final SocketAddress connectAddress;
-        switch (endpoint.getScheme()) {
-          case "tcp":
-            connectAddress = new InetSocketAddress(endpoint.getHostAddress(), endpoint.getPort());
-            break;
-          default:
-            throw new IllegalArgumentException(endpoint.toString());
-        }
+  @Override
+  public void send(@CheckForNull TransportEndpoint endpoint, @CheckForNull Message message) {
+    send(endpoint, message, null);
```

Comment:
Enhancement for the future: I think we need to avoid pass null ever. Let's use similar approach as netty with constant value void promise instead of null.